autometa/common/external/samtools.py

```diff
@@ -42,49 +41,65 @@ def sort(sam, bam, nproc=mp.cpu_count()):
     FileNotFoundError
         Specified path is incorrect or the file is empty
-    ChildProcessError
-        Samtools did not run successfully, returns the return code from subprocessing call
+    subprocess.CalledProcessError
+        Samtools did not run successfully
     """
 
-    if type(nproc) is not int or nproc <= 0:
-        raise TypeError(f"nproc must be an integer greater than zero! Given: {nproc}")
-    if not os.path.exists(sam) or os.stat(sam).st_size == 0:
+    if type(cpus) is not int or cpus <= 0:
+        raise TypeError(f"cpus must be an integer greater than zero! Given: {cpus}")
+    if not os.path.exists(sam) or not os.path.getsize(sam):
         raise FileNotFoundError(
             f"The specified path: {sam} is either incorrect or the file is empty"
         )
-    samtools_out_dir = os.path.dirname(os.path.abspath(bam))
-    err = os.path.join(samtools_out_dir, "samtools.err")
-    out = os.path.join(samtools_out_dir, "samtools.out")
-    with tempfile.TemporaryDirectory() as tempdir:
-        temp_bam = os.path.join(tempdir, os.path.basename(bam))
-        cmd = (
-            f"samtools view -@{nproc} -bS {sam} | samtools sort -@{nproc} -o {temp_bam}"
+    with tempfile.TemporaryDirectory() as tempdir:  # this will delete the temporary files even if program stops in between
+        sort_fpath = os.path.join(tempdir, os.path.basename(bam))
+        # See https://stackoverflow.com/questions/13332268/how-to-use-subprocess-command-with-pipes
+        cmd_view = [
+            "samtools",
+            "view",
+            "-@",
+            str(cpus),
+            "-bS",
+            sam,
+        ]
+        cmd_sort = [
+            "samtools",
+            "sort",
+            "-@",
+            str(cpus),
+            "-o",
+            sort_fpath,
+        ]
+        logger.debug(" ".join(cmd_view))
+        logger.debug(" ".join(cmd_sort))
+        view = subprocess.Popen(
+            cmd_view, stdout=subprocess.PIPE, stderr=subprocess.PIPE
         )
-        logger.debug(f"cmd: {cmd}")
-        with open(out, "w") as stdout, open(err, "w") as stderr:
-            retcode = subprocess.call(cmd, stdout=stdout, stderr=stderr, shell=True)
-        if retcode != 0:
-            raise ChildProcessError(f"Sort failed, retcode: {retcode}")
-        else:
-            shutil.move(temp_bam, bam)
-        os.remove(err)
-        os.remove(out)
-
-
-def main(args):
-    sort(args.sam, args.bam, args.nproc)
+        logger.error(view.stderr)
+        # See https://stackoverflow.com/questions/34147353/python-subprocess-chaining-commands-with-subprocess-run
+        # Popen starts the process and carries on while run starts it and waits for it to finish.
+        sort = subprocess.run(
+            cmd_sort,
+            stdin=view.stdout,
+            stdout=subprocess.DEVNULL,
+            stderr=subprocess.PIPE,
+            check=True,
+        )
+        logger.error(sort.stderr)
+        shutil.move(sort_fpath, bam)
```

Comment:
Why not just remove `check=True` and handle the exception with `sort.check_returncode()`? Or wrap the `subprocess.run` in a `try/except` block?


Could also shorten the stackoverflow link as mentioned by Shaurya. There is a share button that creates a short link to the post.

```suggestion
        logger.error(view.stderr)
        # See https://stackoverflow.com/q/34147353/13118765
        # Popen starts the process and carries on while run starts it and waits for it to finish.
        try:
            sort = subprocess.run(
                cmd_sort,
                stdin=view.stdout,
                stdout=subprocess.DEVNULL,
                stderr=subprocess.PIPE,
                check=True,
            )
        except subprocess.CalledProcessError:
            logger.error(sort.stderr)
        shutil.move(sort_fpath, bam)
```


```suggestion
        logger.error(view.stderr)
        # See https://stackoverflow.com/q/34147353/13118765
        # Popen starts the process and carries on while run starts it and waits for it to finish.
        sort = subprocess.run(
            cmd_sort,
            stdin=view.stdout,
            stdout=subprocess.DEVNULL,
            stderr=subprocess.PIPE,
        )
        if sort.returncode:
            logger.error(sort.stderr)
            return None
        shutil.move(sort_fpath, bam)
```